conn.go

```diff
@@ -39,36 +45,237 @@ type GrpcConnection struct {
 	id            ConnID
 	ip            Address
 	streamWrapper StreamWrapper
+	stopFlag      int32
+	handler       Handler
 	outChan       chan *innerMessage
 	readChan      chan *pb.Message
 	stopChan      chan struct{}
 	sync.RWMutex
 }
 
-func NewConnection(streamWrapper StreamWrapper) (Connection, error) {
-	panic("implement me w/ test case :-)")
+func NewConnection(ip Address, id ConnID, streamWrapper StreamWrapper) (Connection, error) {
+
+	if streamWrapper == nil {
+		return nil, errors.New("fail to create connection ! : streamWrapper is nil")
+	}
+
+	return &GrpcConnection{
+		id:            id,
+		ip:            ip,
+		streamWrapper: streamWrapper,
+		outChan:       make(chan *innerMessage, 200),
+		readChan:      make(chan *pb.Message, 200),
+		stopChan:      make(chan struct{}, 1),
+	}, nil
 }
 
 func (conn *GrpcConnection) Send(data []byte, typ string, successCallBack func(interface{}), errCallBack func(error)) {
-	panic("implement me w/ test case :-)")
+
+	conn.Lock()
+	defer conn.Unlock()
+
+	message, err := conn.build(typ, data)
+	if err != nil {
+		go errCallBack(errors.New(fmt.Sprintf("fail to build message [%s]", err.Error())))
+		return
+	}
+
+	m := &innerMessage{
+		Message:   &message,
+		OnErr:     errCallBack,
+		OnSuccess: successCallBack,
+	}
+
+	conn.outChan <- m
 }
 
-func (conn *GrpcConnection) Ip() string {
-	panic("implement me w/ test case :-)")
+func (conn *GrpcConnection) Ip() Address {
+	return conn.ip
 }
 
-func (conn *GrpcConnection) Id() string {
-	panic("implement me w/ test case :-)")
+func (conn *GrpcConnection) Id() ConnID {
+	return conn.id
 }
 
 func (conn *GrpcConnection) Close() {
-	panic("implement me w/ test case :-)")
+	if conn.isDie() {
+		return
+	}
+
+	isFisrt := atomic.CompareAndSwapInt32(&conn.stopFlag, int32(0), int32(1))
+
+	if !isFisrt {
+		return
+	}
+
+	conn.stopChan <- struct{}{}
+	conn.Lock()
+
+	conn.streamWrapper.Close()
+
+	conn.Unlock()
 }
 
 func (conn *GrpcConnection) Start() error {
-	panic("implement me w/ test case :-)")
+	errChan := make(chan error, 1)
+
+	go conn.readStream(errChan)
+	go conn.writeStream()
+
+	for !conn.isDie() {
+		select {
+		case stop := <-conn.stopChan:
+			conn.stopChan <- stop
+			return nil
+		case err := <-errChan:
+			return err
+		case message := <-conn.readChan:
+			if conn.verify(message) {
+				if conn.handler != nil {
+					m := Message{Message: message, Conn: conn}
+					conn.handler.ServeRequest(m)
+				}
+			} else {
+				// todo: verify 결과 false인 경우
+			}
+		}
+	}
+
+	return nil
+}
+
+func (conn *GrpcConnection) Handle(handler Handler) {
+	conn.handler = handler
 }
 
-func (conn *GrpcConnection) Handle() {
-	panic("implement me w/ test case :-)")
+// TODO : implements me w/ test case
+func (conn *GrpcConnection) verify(envelope *pb.Message) bool {
+	return true
+}
+
+func (conn *GrpcConnection) build(typ string, payload []byte) (pb.Message, error) {
+
+	payloadType, err := messageType(typ, payload)
```

Comment:
I think we can make build message better

## Suggestion
```go
type MessageType interface {
	String() string
	EnumDescriptor() ([]byte, []int)
}
```
`MessageType` is whether sending message is BBA or RBC type. If we declare this information, we can build message much clean

```go
func (conn *GrpcConnection) build(typ MessageType, payload []byte) (pb.Message, error) {
        ...

	message := pb.Message{
		Signature: nil,
		Timestamp: timestamp,
	}
	switch typ.(type) {
	case pb.BBAType:
		message.Payload = &pb.Message_Bba{
			Bba: &pb.BBA{
				Payload:payload,
				Type: typ.(pb.BBAType),
			},
		}
	case pb.RBCType:
            ...
	}


	return message, nil
}
```
In this way, we don't need to know whether sending message is AUX, BVAL, ECHO etc

Reply:
In this code, why did you say `we don't need to know whether sending message is AUX, BVAL, ECHO etc` ?
```go
message.Payload = &pb.Message_Bba{
			Bba: &pb.BBA{
				Payload:payload,
				Type: typ.(pb.BBAType),
			},
		}
```

I think if we use interface, we can distinguish what type of message (ex> in RBC, VAL, ECHO, READY...)

Reply:
> In this code, why did you say we don't need to know whether sending message is AUX, BVAL, ECHO etc ?

because of `messageType` function. In `messageType` function, it checks whether typ is `AUX`, `BVAL` ...
But I think in `Connection`, we don't need to know its concrete type

```
func messageType(typ string, payload []byte) (interface{}, error) {

 	var message interface{}

 	switch typ {
	case "VAL":
		message = pb.Message_Rbc{
			&pb.RBC{
				Payload: payload,
				Type:    pb.RBC_ECHO,
			},
        ...
```

Reply:
Your intention is high layer distinguish message type and just give message to connection. Then I think rbc and bba will distinguish the message type, I removed `build` method and `messageType` function.